stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -1007,3 +1007,124 @@ function matmul3x3!(C::AbstractMatrix, tA, tB, A::AbstractMatrix, B::AbstractMat
     end # inbounds
     C
 end
+
+# Three-argument *
+"""
+    *(A, B::AbstractMatrix, C)
+
+Most 3- and 4-argument `*` calls containing matrices or vectors are done in an
+efficient way, rather than the left-to-right default of `*(x,y,z,...)`.
+
+This can mean performing `B*C` first if `C::AbstractVector`, using five-argument
+`mul!` to fuse the scalar `A::Number` with the matrix multiplication `B*C`,
+or examining `size.((A,B,C,D))` to choose which to multiply first.
+
+!!! compat "Julia 1.6"
+    These optimisations require least Julia 1.6.
+"""
+*(A::AbstractMatrix, B::AbstractMatrix, x::AbstractVector) = A * (B*x)
+
+*(A::AbstractMatrix, x::AbstractVector, γ::Number) = mat_vec_scalar(A,x,γ)
+*(A::AbstractMatrix, B::AbstractMatrix, γ::Number) = mat_mat_scalar(A,B,γ)
+*(α::Union{Real,Complex}, B::AbstractMatrix, C::AbstractVector) = mat_vec_scalar(B,C,α)
+*(α::Union{Real,Complex}, B::AbstractMatrix, C::AbstractMatrix) = mat_mat_scalar(B,C,α)
+
+*(α::Number, u::AbstractVector, tv::AdjOrTransAbsVec) = broadcast(*, α, u, tv)
+*(u::AbstractVector, tv::AdjOrTransAbsVec, γ::Number) = broadcast(*, u, tv, γ)
+*(u::AbstractVector, tv::AdjOrTransAbsVec, C::AbstractMatrix) = u * (tv*C)
+
+*(A::AbstractMatrix, B::AbstractMatrix, C::AbstractMatrix) = _tri_matmul(A,B,C)
+*(tv::AdjOrTransAbsVec, B::AbstractMatrix, C::AbstractMatrix) = (tv*B) * C
+
+function _tri_matmul(A,B,C,δ=nothing)
+    n,m = size(A)
+    # m,k == size(B)
+    k,l = size(C)
+    costAB_C = n*m*k + n*k*l
+    costA_BC = m*k*l + n*m*l
+    if costA_BC < costAB_C
+        isnothing(δ) ? A * (B*C) : A * mat_mat_scalar(B,C,δ)
+    else
+        isnothing(δ) ? (A*B) * C : mat_mat_scalar(A*B, C, δ)
+    end
+end
+
+# Fast path for two arrays * one scalar is opt-in, via mat_vec_scalar and mat_mat_scalar.
+StridedMaybeAdjOrTransMat{T} = Union{StridedMatrix{T}, Adjoint{T, <:StridedMatrix}, Transpose{T, <:StridedMatrix}}
```

Comment:
What was the rationale again here? This doesn't include all the structured `LinearAlgebra` types, right?

Reply:
Oh, okay, I guess these are handled well by the fallback...

Reply:
Correct. I guess something like `_mat_mat_scalar(A::Diagonal, B::Diagonal, γ)` could do better, but not via the existing `mul!` method.